lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -52,86 +60,420 @@
+    // has an associated `@differentiable` attribute.
+    DifferentiableAttribute,
+  };
+
+private:
+  Kind kind;
+  union Value {
+    /// The instruction associated with the `SILSource` case.
+    GradientInst *gradientInst;
+    Value(GradientInst *inst) : gradientInst(inst) {}
+
+    /// The parent differentiation task associated with the
+    /// `IndirectDifferentiation` case.
+    std::pair<ApplyInst *,
+              DifferentiationTask *> indirectDifferentiation;
+    Value(ApplyInst *applyInst, DifferentiationTask *parentTask)
+      : indirectDifferentiation({applyInst, parentTask}) {}
+
+    /// The differential operator associated with the `DifferentialOperator`
+    /// case.
+    ReverseAutoDiffExpr *differentialOperator;
+    Value(ReverseAutoDiffExpr *expr) : differentialOperator(expr) {}
+
+    /// The `@differentiable` attribute associated with the
+    /// `DifferentiableAttribute` case.
+    DifferentiableAttr *differentiableAttribute;
+    Value(DifferentiableAttr *attr) : differentiableAttribute(attr) {}
+  } value;
+
+  /*implicit*/
+  DifferentiationInvoker(Kind kind, Value value)
+    : kind(kind), value(value) {}
+
+public:
+  DifferentiationInvoker(GradientInst *inst)
+    : kind(Kind::GradientInst), value(inst) {}
+  DifferentiationInvoker(ApplyInst *applyInst,
+                         DifferentiationTask *task)
+    : kind(Kind::IndirectDifferentiation), value(applyInst, task) {}
+  DifferentiationInvoker(ReverseAutoDiffExpr *expr)
+    : kind(Kind::DifferentialOperator), value(expr) {}
+  DifferentiationInvoker(DifferentiableAttr *attr)
+    : kind(Kind::DifferentiableAttribute), value(attr) {}
+
+  Kind getKind() const { return kind; }
+
+  GradientInst *getGradientInst() const {
+    assert(kind == Kind::GradientInst);
+    return value.gradientInst;
   }
 
-  SILReverseAutoDiffConfiguration getMasterConfig() const {
-    return SILReverseAutoDiffConfiguration::getMaster(attr->getSourceIndex(),
-                                                      attr->getParamIndices());
+  std::pair<ApplyInst *, DifferentiationTask *>
+  getIndirectDifferentiation() const {
+    assert(kind == Kind::IndirectDifferentiation);
+    return value.indirectDifferentiation;
+  }
+
+  ReverseAutoDiffExpr *getDifferentialOperator() const {
+    assert(kind == Kind::DifferentialOperator);
+    return value.differentialOperator;
+  }
+
+  DifferentiableAttr *getDifferentiableAttribute() const {
+    assert(kind == Kind::DifferentiableAttribute);
+    return value.differentiableAttribute;
   }
+
+  void print(llvm::raw_ostream &os) const;
 };
-} // end anonymous namespace
 
-namespace llvm {
-using swift::SILFunction;
-using swift::SILReverseDifferentiableAttr;
+/// Information about the primal function produced by PrimalGen, e.g.
+/// mappings from the original values to their corresponding ones in the primal
+/// value struct produced by the primal function.
```

Comment:
By "primal value struct", you must be referring to the struct containing primal intermediate values.
An explanation of the purpose of the primal value struct would be nice, e.g. "the primal value struct stores intermediate values from the original function".

Reply:
Added explanation.